packages/yarnpkg-pnpify/sources/WatchManager.ts

```diff
@@ -0,0 +1,77 @@
+import {Filename, PortablePath, Watcher, WatchCallback} from '@yarnpkg/fslib';
+
+import {EventEmitter}                                   from 'events';
+
+import {NodePathResolver}                               from './NodePathResolver';
+
+class WatchEventEmitter extends EventEmitter {
+  private dirWatchers: DirectoryWatcherMap;
+  private watchPath: PortablePath;
+  private watcherId: number;
+
+  public constructor(dirWatchers: DirectoryWatcherMap, watchPath: PortablePath, watcherId: number) {
+    super();
+    this.dirWatchers = dirWatchers;
+    this.watchPath = watchPath;
+    this.watcherId = watcherId;
+  }
+
+  public close() {
+    const dirWatcher = this.dirWatchers.get(this.watchPath)!;
+    dirWatcher.eventEmitters.delete(this.watcherId);
+    if (dirWatcher.eventEmitters.size === 0) {
+      this.dirWatchers.delete(this.watchPath);
+    }
+  }
+}
+
+type DirectoryWatcherMap = Map<PortablePath, DirectoryWatcher>;
+
+interface DirectoryWatcher {
+  lastWatcherId: number;
+  eventEmitters: Map<number, Watcher & EventEmitter>;
+  dirEntries: Set<Filename>;
+}
+
+export class WatchManager extends EventEmitter {
+  private readonly dirWatchers: DirectoryWatcherMap = new Map();
+
+  public registerWatcher(watchPath: PortablePath, dirList: Set<Filename>, callback: WatchCallback) {
+    let dirWatcher = this.dirWatchers.get(watchPath);
+    if (!dirWatcher) {
+      dirWatcher = {lastWatcherId: 0, eventEmitters: new Map(), dirEntries: dirList};
+      this.dirWatchers.set(watchPath, dirWatcher);
+    }
+    const watcherId = dirWatcher.lastWatcherId++;
```

Comment:
We don't actually need to keep `lastWatcherId` in each `DirectoryWatcher`, right? We could keep a single one in the `WatchManager` and use it everywhere (since `eventEmitters` is a map anyway)